src/proxy-animation.js

```diff
@@ -30,151 +30,470 @@ function createReadyPromise(details) {
+      updateFinishedState(proxyAnimations.get(this));
+    }
+  }
+
+  // -----------------------------------------
+  // Web animation API
+  // -----------------------------------------
+
   get effect() {
     return proxyAnimations.get(this).animation.effect;
   }
   set effect(newEffect) {
-    return proxyAnimations.get(this).animation.effect = newEffect;
+    proxyAnimations.get(this).animation.effect = newEffect;
   }
 
   get timeline() {
-    const internalTimeline = proxyAnimations.get(this).timeline;
-    if (internalTimeline !== undefined)
-      return internalTimeline;
-    return proxyAnimations.get(this).animation.timeline;
+    const details = proxyAnimations.get(this);
+    return details.timeline || details.animation.timeline;
+  }
+  set timeline(newTimeline) {
+    const oldTimeline = this.timeline;
+    if (oldTimeline == newTimeline)
+      return;
+
+    const details = proxyAnimations.get(this);
+
+    const fromScrollTimeline = (oldTimeline instanceof ScrollTimeline);
+    const toScrollTimeline = (newTimeline instanceof ScrollTimeline);
+    const previousCurrentTime = this.currentTime;
+    const previousPlayState = this.playState;
+    const playbackRate = effectivePlaybackRate(details);
+    const pending = this.pending;
+
+    if (fromScrollTimeline) {
+      removeAnimation(details.timeline, details.animation);
+    }
+
+    details.resetCurrentTimeOnResume = false;
+    if (toScrollTimeline) {
+      // Cannot assume that the underlying native implementation supports
+      // mutable timelines. Thus, we leave its timeline untouched, and simply
+      // ensure that it is in the paused state.
+      details.timeline = newTimeline;
+      applyPendingPlaybackRate(details);
+      details.animation.pause();
+      switch(previousPlayState) {
+        case 'idle':
+          details.playState = 'idle';
+          details.holdTime = null;
+          details.startTime = null;
+          break;
+
+        case 'paused':
+          details.playState = 'paused';
+          details.resetCurrentTimeOnResume = true;
+          details.animation.currentTime = previousCurrentTime;
+          break;
+
+        case 'running':
+        case 'finished':
+          details.playState = 'running';
+          details.startTime =
+              playbackRate < 0 ? details.animation.effect.getTiming().duration
+                               : 0;
+          details.holdTime = null;
+          break;
+      }
+      addAnimation(details.timeline, this);
+      if (pending)
+        createReadyPromise(details);
+    } else {
+      // TODO: polyfill mutable timeline support. Cannot assume the native
+      // animation supports mutable timelines. Could keep a list of detached
+      // timelines and pump updates to current time via rAF.
+      details.animation.timeline = newTimeline;
+      if (fromScrollTimeline) {
+        // TODO: sync pending status & play state (ready promise).
+        details.timeline = null;
```

Comment:
I think we should do this earlier, perhaps first thing in the else branch? If the assignment on 242 throws we might not run this right?